### ISSUE TITLE
Support views

### ISSUE THREAD

**Issue by grosch** [vapor/fluent#493]
Please provide a way to create and query against views instead of just tables.  For example, here's a view that I currently have in my database that gets used quite a bit.

```sql
CREATE VIEW public.v_work_completed AS
SELECT w.id, w.date, w.amount, pt.amount "value", w.amount * pt.amount total, p.name, pt.abbrev, o.id order_id
FROM work_completed w
INNER JOIN producers p ON w.producer_id = p.id
INNER JOIN order_points op ON w.order_point_id = op.id
INNER JOIN points pt ON op.points_id = pt.id
INNER JOIN orders o ON op.order_id = o.id;
```

Getting that from the tables would be a huge amount of repetitive code in Swift.  

I'm not looking to be able to do inserts through the view, just have the view created, and then be able to do read operations against it.

**Comment by MrMage** [comment]
I’m not an expert, but wouldn’t it already be possible to create a Model subclass with the view‘s name as its table name, then ru  Fluent queries against that model? (View creation would not be covered, of course.)

**Comment by grosch** [comment]
Testing now...

**Comment by grosch** [comment]
No, I don't think that'll work.  If I don't add it as a Migration it seems to just get ignored.  And if I try to add it as a migration it wants default values for everything to be set.

**Comment by grosch** [comment]
And if I do give default values so it compiles, then the "migration" fails as the view already exists and it's trying to create it.

**Comment by MrMage** [comment]
I was only talking about fetching, not about creating it. As I said, View creation would not be covered.

If you create a model with the view‘s name as table name, then use Fluent‘s fetch methods?

**Comment by grosch** [comment]
It didn't seem to do anything when I tried that.

**Comment by MrMage** [comment]
What did you try, and what SQL code got executed by that?

**Comment by grosch** [comment]
I created the model like so:

```swift
import FluentPostgreSQL
import Vapor

final class ViewWorkCompleted: Codable {
    static let entity = "v_work_completed"

    var id: Int? = nil
    var date = Date()
    var amount: Double = 0
    var value: Double = 0
    var total: Double = 0
    var name: String = ""
    var abbrev: String = ""
    var orderID: Order.ID = 0

    enum CodingKeys: String, CodingKey {
        case id, date, amount, value, total, name, abbrev
        case orderID = "order_id"
    }
}

extension ViewWorkCompleted: PostgreSQLModel {}
```

And then I just tried to query everything 

```swift
_ = ViewWorkCompleted.query(on: req).all().map(to: Void.self) { rows in print(rows) }
```

There are 24k rows in that view's output if I directly query the database.

I don't know how to tell it to print out the SQL that it ran.

**Comment by MrMage** [comment]
Make sure to wait on the result somewhere (for testing purposes!) to ensure the query actually gets executed and/or receive the error it returned.

Also, see https://github.com/vapor/database-kit/blob/0db303439e5ef8b6df50a2b6c4029edddee90cb0/Sources/DatabaseKit/Database/DatabaseConfig.swift#L67 on enabling logging.

**Comment by grosch** [comment]
How do I get a database identifier?  I setup my db by just registering a PostgreSQLDatabaseConfig.

**Comment by MrMage** [comment]
https://github.com/vapor/postgresql/blob/9721627e9dfa103cc41abd8e4a7e3a937fa6cb63/Sources/PostgreSQL/Database/PostgreSQLDatabase.swift#L37

**Comment by grosch** [comment]
Could you please provide a small code sample.  This is hard to chase this stuff down.  I tried this:

```swift
    services.register(PostgreSQLDatabase(config: dbConfig))
    services.register(dbConfig)

    var dbkit = DatabasesConfig()
    dbkit.enableLogging(on: DatabaseIdentifier<PostgreSQLDatabase>.psql)
    services.register(dbkit)
```

but when I run it fails saying:
 
> [DatabaseKitError.dbRequired: No database with id 'psql' is configured.]

**Comment by grosch** [comment]
OK, finally got it right:

```swift
    var dbkit = DatabasesConfig()
    let db = PostgreSQLDatabase(config: dbConfig)
    dbkit.add(database: db, as: DatabaseIdentifier<PostgreSQLDatabase>.psql)
    dbkit.enableLogging(on: DatabaseIdentifier<PostgreSQLDatabase>.psql)
    services.register(dbkit)
```

It doesn't print anything out related to the `ViewWorkCompleted` query, though I do see all the others.

**Comment by MrMage** [comment]
Could you please provide more context around the `_ = ViewWorkCompleted.query(on: req).all().map(to: Void.self) { rows in print(rows) }` call?

**Comment by grosch** [comment]
What would you like to know?  I have that inside of one of my routes:

```swift
    func activeOrderIndexHandler(_ req: Request) throws -> Future<View> {
        let month = startOfMonth()

        _ = ViewWorkCompleted.query(on: req).all().map(to: Void.self) { rows in print(rows) }

        return try Order.query(on: req)
            .filter(\Order.month == month)
            .all()
            .flatMap(to: View.self) { orders in
                let content = orders.map { OrderViewModel(order: $0, req: req) }
                let context = OrderIndexContext(orders: content.isEmpty ? nil : content)

                return try req.renderLeaf("orders/index", context)
        }
    }
```

And when I run I see this in the output:

> [psql] [2018-05-26 08:05:05 +0000] select oid, relname from pg_class []
[psql] [2018-05-26 08:05:05 +0000] SELECT * FROM "orders" WHERE ("orders"."month" = $1) ["TIMESTAMP (binary) \0\u{02}\u{0E}\u{18}k \0)"]
[psql] [2018-05-26 08:05:06 +0000] SELECT * FROM "points" JOIN "order_points" ON "points"."id" = "order_points"."points_id" WHERE ("order_points"."order_id" = $1) ["BIGINT (binary) \0\0\0\0\0\0+#)"]

The first line is fluent of course, the second one is coming from that `Order.query` I do there, and then that last one repeats tons of times as it's called inside of the `OrderViewModel` that's looped.
 
I would have expected to see a line like:

> SELECT * FROM v_work_completed

if it were doing anything.

**Comment by grosch** [comment]
It's 1am, I gotta head to bed.  Maybe you can just do a quick test yourself to see what's going on?

**Comment by MrMage** [comment]
Try what this prints:

        return ViewWorkCompleted.query(on: req).all().flatMap(to: View.self) { rows in
            print(rows)
            return try Order.query(on: req)
            .filter(\Order.month == month)
            .all()
            .flatMap(to: View.self) { orders in
                let content = orders.map { OrderViewModel(order: $0, req: req) }
                let context = OrderIndexContext(orders: content.isEmpty ? nil : content)

                return try req.renderLeaf("orders/index", context)
        } }

**Comment by MrMage** [comment]
> It's 1am, I gotta head to bed. Maybe you can just do a quick test yourself to see what's going on?

Sorry, but I have a weekend, too.

**Comment by grosch** [comment]
[ ERROR ] VaporError.defaultDB: `Model.defaultDatabase` is required to use request as `DatabaseConnectable`. (Request.swift:161)
[ DEBUG ] Suggested fixes for VaporError.defaultDB: Ensure you are using the 'model' label when registering this model to your migration config (if it is a migration): migrations.add(model: ..., database: ...). If the model you are using is not a migration, set the static `defaultDatabase` property manually in your app's configuration section. Use `req.withPooledConnection(to: ...) { ... }` instead. (Logger+LogError.swift:20)

**Comment by MrMage** [comment]
Did you try setting `Model.defaultDatabase = ...` when configuring your application object, as suggested by this?

**Comment by grosch** [comment]
No because I don't really know what that means.  It doesn't tell me what class that static property exists on, and my ViewWorkCompleted model doesn't have defaultDatabase as a static property that it inherited.

**Comment by MrMage** [comment]
It's a static property on `Model` itself, and it needs to be set to the same database identifier as mentioned before.

**Comment by grosch** [comment]
`Model.defaultDatabase = DatabaseIdentifier<PostgreSQLDatabase>.psql`

If I try that it says

> Member 'defaultDatabase' cannot be used on value of protocol type 'Model'; use a generic constraint isntead
> Protocol 'Model' can only be used as a generic constraint because it has Self or associated type requirements

**Comment by grosch** [comment]
Oh, duh.  ViewWorkCompleted.defaultDatabase = ...

:)

**Comment by MrMage** [comment]
Sorry, I mistook the two as well. What are you getting now?

**Comment by grosch** [comment]
It's showing me a huge array of ViewWorkCompleted!  I think it might be working now.  I've gotta change some things around to try and use it.  I'll play for a while. 

Thank you SO much for sticking with me on this.  I'm new to Fluent so everything that should be simple is still a learning curve for me.

**Comment by grosch** [comment]
This is working perfectly, thanks again.  I'm going to close this one and open a new issue that's just explicitly a feature request to add creating a view.

**Comment by MrMage** [comment]
Glad to hear that! I agree that the documentation is sometimes thin, so it’s hard to look things up. In those cases, I like to use Xcode‘s „find in project“ feature, looking where the type or method I’m looking for is being declared or used. That has really helped me find many things and how to use them.

**Comment by chrigu1981** [comment]
just add this to your configure.swift file where User is one of your Models
```User.defaultDatabase = .mysql```
